File: src/services/ranking/__tests__/performanceGrade.test.ts
import { performanceGrade, performanceScore } from '@/services/ranking/performanceGrade';

describe('performanceScore (today vs baseline)', () => {
  it('is neutral with no baseline or no valid effort (new users not punished)', () => {
    expect(performanceScore(120, null)).toBe(60);
    expect(performanceScore(null, 100)).toBe(60);
  });
  it('scores at/above and below the baseline', () => {
    expect(performanceScore(100, 100)).toBeCloseTo(65, 5); // ratio 1.0
    expect(performanceScore(110, 100)).toBeCloseTo(88, 5); // ratio 1.1
    expect(performanceScore(90, 100)).toBeCloseTo(40, 5); // ratio 0.9
  });
  it('maps to a grade letter', () => {
    expect(performanceGrade(performanceScore(100, 100))).toBe('A'); // 65 -> A
    expect(performanceGrade(performanceScore(90, 100))).toBe('C'); // 40 -> C
  });
});

File: android-native/app/src/test/java/com/fitnessrpg/app/domain/ranking/PerformanceGradeTest.kt
package com.fitnessrpg.app.domain.ranking

import org.junit.Assert.assertEquals
import org.junit.Test

class PerformanceGradeTest {

    @Test
    fun `is unscored with no baseline or no valid effort`() {
        assertEquals(0.0, performanceScore(120.0, null), 1e-9)
        assertEquals(0.0, performanceScore(null, 100.0), 1e-9)
    }

    @Test
    fun `scores at, above, and below the baseline`() {
        assertEquals(65.0, performanceScore(100.0, 100.0), 1e-4) // ratio 1.0
        assertEquals(88.0, performanceScore(110.0, 100.0), 1e-4) // ratio 1.1
        assertEquals(40.0, performanceScore(90.0, 100.0), 1e-4) // ratio 0.9
    }

    @Test
    fun `uses words and never a temporary rank letter`() {
        assertEquals("Baseline", todayPerformanceLabel(100.0, null))
        assertEquals("Normal", todayPerformanceLabel(100.0, 100.0))
        assertEquals("PR", todayPerformanceLabel(110.0, 100.0, isPr = true))
    }
}
